### ISSUE TITLE
Replace VirtualServices for egress traffic by ServiceEntry Endpoints

### ISSUE THREAD

**Issue by vadimeisenbergibm** [istio/istio.io#3312]
Direct traffic thru egress gateway by using endpoints:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: google
spec:
  hosts:
  - www.google.com
  ports:
  - number: 443
    name: https
    protocol: HTTPS
  resolution: DNS
  endpoints:
  - address: www.google.com
    network: external
  - address: istio-egressgateway.istio-system.svc.cluster.local
```

**Comment by vadimeisenbergibm** [comment]
@ibm-jason Could you please handle this is issue?
cc @linsun @rshriram

**Comment by ibm-jason** [comment]
@vadimeisenbergibm sure thing.

**Comment by ibm-jason** [comment]
/cc @vadimeisenbergibm 

I've taken a look at the documentation and identified these areas where VirtualServices have been defined for egress traffic:

- https://preliminary.istio.io/docs/examples/advanced-gateways/egress-gateway/#egress-gateway-for-http-traffic
  - Egress gateway for HTTP traffic, Step 4
  - Egress gateway for HTTPS traffic, Step 3
- https://preliminary.istio.io/docs/examples/advanced-gateways/egress-gateway-tls-origination/
  - Perform TLS origination with an egress gateway, Step 4
  - Deploy a mutual TLS server, Step 6
  - Configure mutual TLS origination for egress traffic, Step 2

Are there any other places in the documentation that I haven't identified?

**Comment by vadimeisenbergibm** [comment]
@ibm-jason Note that VirtualServices are redundant only if they do not rewrite the port.

In https://preliminary.istio.io/docs/examples/advanced-gateways/egress-gateway-tls-origination/
* Perform TLS origination with an egress gateway, Step 4 - rewrite port 80 to 443

**Comment by andraxylia** [comment]
@ibm-jason I maked this as a P0, please let me know if you are working on this for 1.1.

**Comment by ibm-jason** [comment]
@andraxylia Yes, I am working on this for 1.1.

**Comment by vadimeisenbergibm** [comment]
@rshriram @ibm-jason @frankbu @esnible @linsun I have reservations regarding this issue. Let me explain the situation:

There are two incompatible types of egress gateways, the old one and the new one. The new one has `ISTIO_META_REQUESTED_NETWORK_VIEW: “external”` environment variable set. The old one is configured by ServiceEntry + VirtualServices, and most of the examples with Egress Gateway use the old egress gateway. The new one is configured by ServiceEntry with endpoints, the multicluster example with gateway connectivity uses the new one, see https://preliminary.istio.io/docs/examples/multicluster/gateways/#send-remote-cluster-traffic-using-egress-gateway. The old gateway is deployed by using values.yaml, the new gateway is deployed by using values-istio-multicluster-gateways.yaml.

The two different styles of the gateways will not understand configurations of one another and will fail if configured in a different style. Using two different gateway styles can cause users confusion and problems with configuration, not working as expected. 

I propose the following:
1. To a use a single style of egress gateway, both for multicluster and for egress traffic control to external services. It is now possible, especially since now it is possible to direct multicluster traffic through 443 port (I work on https://github.com/istio/istio/issues/12224).
2. To check if we can switch to the new gateway style in all the egress examples. If we cannot, stick to the old style.
3. If we can switch to the new gateway style in all the examples now, consider to switch to it in 1.2 and not in 1.1.

**Comment by linsun** [comment]
+1 on switch to the new gateway style.  our egress gateway is confusing enough - anything that can simplify it, i'm all in favor of it.

**Comment by linsun** [comment]
for 1.1, i think that is p0 is that  we have docs updated for this - "vs is no longer needed when using service entry to allow outbound traffic".  cc @andraxylia

**Comment by vadimeisenbergibm** [comment]
@ibm-jason Please check that the endpoints can be used in *all the cases*. Please remember to use the `ISTIO_META_REQUESTED_NETWORK_VIEW: “external”` annotation in the deployment of the egress gateway. To set the annotation, use the following values snippet:
```
gateways:
  istio-egressgateway:
    env:
      # Needed to route traffic via egress gateway if desired.
      ISTIO_META_REQUESTED_NETWORK_VIEW: "external"
```

or set the value of `gateways.istio-egressgateway.env.ISTIO_META_REQUESTED_NETWORK_VIEW` in `helm template`.

**Comment by geeknoid** [comment]
This is a P0 blocking the 1.1 release. When you believe this issue can be addressed? If you don't have time, please assign to someone else that can address this issue.

Thanks.

**Comment by geeknoid** [comment]
@vadimeisenbergibm @ibm-jason @linsun @frankbu @rshriram 

Can someone take a look at this issue please? This is blocking the 1.1 release

**Comment by ibm-jason** [comment]
@vadimeisenbergibm @linsun I'm going to reach out to you about this.

**Comment by frankbu** [comment]
@ibm-jason have you been able to check if all of @vadimeisenbergibm examples still work if you configure the gateway  with `ISTIO_META_REQUESTED_NETWORK_VIEW: "external"? If so, then I think this is just a simple one line change to turn on the new style in the command for starting the gateway. Right?

Otherwise, I agree that if all of his examples can not be made to work with the new style egress gateway configured with `ISTIO_META_REQUESTED_NETWORK_VIEW: "external"`, then we should just leave them as they are for now, and switch later. I agree that using a mixture of the two styles will likely cause problems. In this case, I think @vadimeisenbergibm is suggesting that we should change the the multicluster gateways instruction to use the old style, right?

If we leave everything as is for now, is it really so bad? We could change it in a point release too (1.1.1), right @vadimeisenbergibm ?

**Comment by ibm-jason** [comment]
@frankbu I think you're correct, but I've yet to check all the examples, and this is what I wanted to chat with Vadim about.

**Comment by vadimeisenbergibm** [comment]
@frankbu I agree with you, except for port rewriting - it seems now that port rewriting can be done with the new style too. So it seems everything can be done with the new style, but it has to be verified.

**Comment by linsun** [comment]
@ibm-jason can you pls update which egress page you have validated and which are remaining?

**Comment by vadimeisenbergibm** [comment]
Blocked by https://github.com/istio/istio/issues/12522.

**Comment by ibm-jason** [comment]
@linsun To summarize, @vadimeisenbergibm and I worked on creating a EgressGateway using a ServiceEntry definition that uses an endpoint, and we found that testing it via the `curl` command in the documentation returns a `404` response. See istio/istio#12522 for details.

**Comment by geeknoid** [comment]
So what should we do? Can we lower this to P1?

**Comment by vadimeisenbergibm** [comment]
@geeknoid, @linsun told us that its priority is lowered to P1 already.

**Comment by ibm-jason** [comment]
## Status Update

While retesting all of these examples using Service Entry endpoints, I discovered that the custom resource definitions for each of the step are malformed. When copying/pasting the CRDs in Step 3 for the "Configure an Egress Gateway" example, the command returns the following error:

`error: error parsing STDIN: error converting YAML to JSON: yaml: line 8: mapping values are not allowed in this context`

Talked with @frankbu about this issue, who opened https://github.com/istio/istio.io/pull/3847 to resolved the problem.
